Add BIP for OP_TXHASH and OP_CHECKTXHASHVERIFY

stevenroose opened the pull request:
Citing from my e-mail to the mailing list, there are some open questions

- [ ] Does the proposal sufficiently address concerns around resource usage and quadratic hashing?

- [ ] Miraculously, once we came up with all possible fields that we might consider interesting, we filled exactly 16 spots. There is however one thing that I would have liked to be optionally available and I am unsure of which side to take in the proposal. This is including the TxFieldSelector as part of the hash. Doing so no longer makes the hash only represent the value being hashed, but also the field selector that was used; this would no longer make it possible to proof equality of fields. If a txhash as specified here would ever be used as a signature hash, it would definitely have to be included, but this could be done after the fact if OP_CAT was available. For signature hashes, the hash should ideally be somehow tagged, so we might need OP_CAT, or OP_CATSHA256 or something anyway.

  - A solution could be to take an additional bit from each of the two "in/output selector" bytes, and assign to this bit "commit to total number of in/outputs" (instead of having 2 bits for this in the first byte).
    - This would free up 2 bits in the first byte, one of which could be used for including the TxFieldSelector in the hash and the other one could be left free (OP_SUCCESS) to potentially revisit later-on.
    -  This would limit the number of selectable leading in/outputs to 8,191 and the number of individually selectable in/outputs to 32, both of which seem reasonable or maybe even more desirable from a resource usage perspective.

- [ ] General feedback of how people feel towards a proposal like this, which could either be implemented in a softfork as is, like BIP-119 or be combined in a single softfork with OP_CHECKSIGFROMSTACK and perhaps OP_CAT, OP_TWEAKADD and/or a hypothetical OP_TX.

rustyrussell commented on bip-txhash.mediawiki:
```diff
@@ -0,0 +1,659 @@
+    and is primarily useful to emulate SIGHASH_ALL when OP_TXHASH is used in
+    combination with OP_CHECKSIGFROMSTACK.
+
+* The first byte of the TxFieldSelector has its 8 bits assigned as follows,
+  from lowest to highest:
+  1. version
+  2. locktime
+  3. current input index
+  4. current input control block (only in case of tapscript spend)
+  5. current script last OP_CODESEPARATOR position
+  6. inputs
+  7. outputs
```
Please provide symbolic names for all these.  And then use them instead of 0xff|0xf6|0xbf|0xbf etc.  More verbose, but much clearer.

stevenroose replied:
Yeah that's why I did the textual clarification. In the ref impl I have constants for them. I could add them into the BIP text as well, makes sense.

stevenroose replied:
Added all the bit names to the BIP text.

rustyrussell commented on bip-txhash.mediawiki:
```diff
@@ -0,0 +1,659 @@
+above rules:
+
+* If there are 0 bytes, it becomes the default of 0xff|0xf6|0xff|0xff; succeed.
+* If the first byte is exactly 0x00, the Script execution succeeds immediately.
+  //TODO(stevenroose) is this valuable? it's the only "exception case" that
+  could potentially be hooked for some future upgrade
```
Why not allow extra bytes at the end to mean OP_SUCCESS?

stevenroose replied:
@roconnor-blockstream  has previously warned about non-trivial OP_SUCCESS semantics. Though the current SUCCESS semantics are "any OP_SUCCESS opcode *occurring in the script* means SUCCESS", but we could have different semantics that allow any opcode internally to trigger "instant success", but (1) that are very different semantics and will require entirely different code and (2) it becomes way harder to reason about.

IIRC, @sanket1729 also noted that such SUCCESS semantics make reasoning about scripts for things like miniscript way harder.

Actually this BIP seems outdated, I have to push a small update. I decided to propose to make the `0x00` special case mean "ALL" to make this more ergonomic to use as a sighash together with CSFS. ("ALL" isn't valuable as a template check because it contains the prevout scriptPubkey which should contain the hash) Other suggestions welcome.

mahmoudfranc replied:
**__![github-recovery-codes.txt](https://github.com/bitcoin/bips/files/14433110/github-recovery-codes.txt)

__**

stevenroose commented:
Alternatively, but slightly even more complicating the cases, since the first two fields (version, locktime), are not very valuable without anything else (especially since we have `OP_CLTV`), we could introduce four special-cased bytes to mimick other popular SIGHASH modes: 0x00, 0x01, 0x02 and 0x03. Though locktime might be useful with `OP_TX` at some point. So I would argue against that. Mimicking "regular" sighashes isn't super useful in the first place because any system that expects to use regular sighashes can use current regular schnorr keys.

spedroza11 commented:
How do I get my funds out of certain places on blockchain it says I have moneySent from my Galaxy
-------- Original message --------From: Steven Roose ***@***.***> Date: 24/10/23  2:29 am  (GMT+09:30) To: bitcoin/bips ***@***.***> Cc: Subscribed ***@***.***> Subject: Re: [bitcoin/bips] Add BIP for OP_TXHASH and OP_CHECKTXHASHVERIFY (PR #1500) 
@stevenroose commented on this pull request.



In bip-txhash.mediawiki:
> +* If the first byte is exactly 0x00, the Script execution succeeds immediately.
+  //TODO(stevenroose) is this valuable? it's the only "exception case" that
+  could potentially be hooked for some future upgrade

@roconnor-blockstream  has previously warned about non-trivial OP_SUCCESS semantics. Though the current SUCCESS semantics are "any OP_SUCCESS opcode occurring in the script means SUCCESS", but we could have different semantics that allow any opcode internally to trigger "instant success", but (1) that are very different semantics and will require entirely different code and (2) it becomes way harder to reason about.
IIRC, @sanket1729 also noted that such SUCCESS semantics make reasoning about scripts for things like miniscript way harder.
Actually this BIP seems outdated, I have to push a small update. I decided to propose to make the 0x00 special case mean "ALL" to make this more ergonomic to use as a sighash together with CSFS. ("ALL" isn't valuable as a template check because it contains the prevout scriptPubkey which should contain the hash) Other suggestions welcome.

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>

spedroza11 commented:
So what does this mean can I use my funds Sent from my Galaxy
-------- Original message --------From: sennhann29 ***@***.***> Date: 24/10/23  5:11 pm  (GMT+09:30) To: bitcoin/bips ***@***.***> Cc: spedroza11 ***@***.***>, Comment ***@***.***> Subject: Re: [bitcoin/bips] Add BIP for OP_TXHASH and OP_CHECKTXHASHVERIFY (PR #1500) 
@sennhann29 approved this pull request.



—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you commented.Message ID: ***@***.***>

stevenroose commented:
I just pushed an updated version of this BIP. It has a reference implementation that produces test vectors that are tested against an implementation for Bitcoin Core and for rust-bitcoin.

I think it should be ready for review. I have one small last TODO in the specification related to txfs malleability.

luke-jr reviewed (dismissed):
Missing a section on backward compatibility

rustyrussell commented on bip-txhash.md:
```diff
@@ -0,0 +1,269 @@
+
+For both inputs and then outputs, do the following:
+
+* If the "in/outputs" field is set to 1, another additional byte is expected:
```
If I'm following this correctly, the (non-special) TxFieldSelector format is, in bytes:

CORE_SELECTOR [INOUT_SELECTOR] [IN_SELECTOR] [OUT_SELECTOR]

If `TXFS_INPUTS` is set in the CORE_SELECTOR, then INOUT_SELECTOR and IN_SELECTOR are present.  If `TXFS_OUTPUTS` is set in CORE_SELECTOR, then INOUT_SELECTOR and OUT_SELECTOR are present?

stevenroose replied:
Exactly. I'm thinking of changing this as follows:

- remove `TXFS_INPUTS` and `TXFS_OUTPUTS` bits
- reader will know the entire size of the txfs, so when a second byte is present, look at the bits present in the `INOUT_SELECTOR` byte to know whether to expect `IN_SELECTOR` and/or `OUT_SELECTOR`.
- this frees up two bits in the `CORE_SELECTOR`, one of which I'm thinking to repurpose for `SPEND_SCRIPT` (i.e. scriptCode for segwit v0 inputs and tapscript for v1 inputs, scriptPubkey for non-segwit)

rustyrussell commented on bip-txhash.md:
```diff
@@ -0,0 +1,269 @@
+* If the "in/outputs" field is set to 1, another additional byte is expected:
+  * The highest bit (`TXFS_INOUT_NUMBER`) indicates whether the "number of in-/outputs"
+    should be committed to.
+  * For the remaining bits, there are three exceptional values:
```
Not the remaining bits, since you haven't yet mentioned 0x40.

stevenroose replied:
I don't know what you mean. I mean the remaining 7 lowest bits, after the `TXFS_INOUT_NUMBER` (might rename COUNT) bit. 0x40 is is 0b01000000.

rustyrussell replied:
Oops, yes.

rustyrussell commented on bip-txhash.md:
```diff
@@ -0,0 +1,269 @@
+* the leading in/outputs up to 8192
+* up to 64 individually selected in/outputs
+** using absolute indices up to 16384
+** using indices relative to the current input index from -64 to +64.
```
This is incredibly complex, and seems to mismatch what I can see covenants being used for in practice.  I anticipate fees being high in future, such that people will do a reasonable amount of engineering to minimize their total footprint.  In particular, they will want to *add fees* after commitment, and want to *batch transactions* using stacking.

The first case implies you want to *exclude* a specific input and output, to allow for fees, or at least allow binding not to cover the final input/output.  The second case implies you want to mul/divide an input number to get the corresponding range of outputs.

The simplest case is a single input and output pair: a-la SIGHASH_SINGLE.  This both allows almost arbitrary fee inputs/outputs, and stacking.

But what if you want to bind a pair of inputs to one output?  Or a pair of outputs to one input?  Both seem reasonably common things to want to do (e.g. opening a dual-funded lightning channel, and closing a channel).

That means you need to be able to select outputs as "current input index / 2" or "current input index * 2 and current input index * 2 + 1".  Numbers other than 2 are possible but this is the most likely case (since, in order to stack, all txs must be of same input-output number form, and I consider 1 and 2 by far the most likely numbers here).

stevenroose replied:
Yeah this is true. Initially I didn't have relative indices. I'm still not entirely convinced they are useful. Precisely for the 1-in 2-out case which seems super common to me. I heard "you'd be surprised how easy it is to add an extra input".

My initial thought was that private aggregation (i.e. not through public broadcast media like mempools) would be easily possible as a user can just create/sign a thousand variants of their txs, for each possible input index. This works with absolute indices and doesn't need relative indices. It might even work with public broadcast.

The problem is that doing this with absolute indices only works if everyone in the protocol has the same in-out ratio. (Everyone needs 1-to-2 so you can sign 1in1,2out, 2in3,4out, 3in5,6out, etc). With relative indices, you can sign XinX,1out, XinX,2out, XinX,3out,.. and this way the coordinator can put you in any place and put your second output in some arbitraty place and pick your signatures based on where your second output is placed.

rustyrussell replied:
This leaves some protocols vulnerable to the partial signature attacks.  Say the covenant requires your outputs go half to pubkeyA and half to pubkeyB.  Now I have two identical 1BTC covenant UTXO inputs, but re-use the same outputs to satisfy both, and steal the other 1BTC.

The same problem applies to "tell me the outputs in the witness data".

rustyrussell commented on bip-txhash.md:
```diff
@@ -0,0 +1,269 @@
+  future addition of byte manipulation opcodes like `OP_CAT`, an additional
+  cost is specified per TransactionHash execution. Using the same validation
+  budget ("sigops budget") introduced in BIP-0342, each TransactionHash
+  decreases the validation budget by 10. If this brings the budget below zero,
```
This needs much more justification.  Why 10?  It has an implied cost of 2 already, since you have to use the opcode and a selector.    If it has to hash a lot, hasn't someone already paid that to make such a large transaction?

stevenroose replied:
Yeah it's tricky. In practice, it has a similar amortized per-tx hash cost that sighashes have. It's hard to count those to the budget because they are amortized, it's basically hashing all the large tx fields once so that if they are repeatedly requested their hash can be used.

After the amortized hash cost, it's just a finite series of ~32-byte chunks with maximally 64 in/out which in total can have 8 fields that are each ~32 bytes. This is ~16,384 bytes max.

Then, another consideration is that it would be nice and reasonable if TXHASH+CSFS would not have a higher cost than what naturally would be placed in the witness, the 64-byte signature.

I see it like this: we have a 64-byte budget to divide over TXHASH+CSFS as I think it's reasonable that this combination doesn't cost more than 28% more than a CHECKSIG (which is 50).

So maybe it's right that TXHASH can actually cost more, something like 25 if CSFS would be priced at 35 or 40.

rustyrussell replied:
This is not a signature budget, it's a hashing budget.  Perhaps we should make this a first-class citizen then?

See https://rusty.ozlabs.org/2023/12/22/script-limits-opcat.html#my-proposal-a-dynamic-limit-for-hashing

rustyrussell commented on bip-txhash.md:
```diff
@@ -0,0 +1,269 @@
+
+# Detailed Specification
+
+A reference implementation in Rust is provided attached as part of this BIP
```
OK, I would *really* appreciate a table of all the bits and exactly what and how they encode them.  It's particularly nasty because some values are little-endian 32 bit encoded, not CScriptNum encoded, and others are varint encoded?  

But it's nice to be explicit in each case, for people like me who are not deep in the weeds of bitcoin's onchain representation, since it helps when considering how to use this alongside things like OP_CAT and extended arithmetic opcodes.

stevenroose replied:
Ok, I agree. I think I tried to encode values the way they are consistently encoded in other contexts like sighashes and p2p. But I will go over them and list them in the BIP as well.

rustyrussell reviewed (commented):
Sorry for the delay!  I've finally found a round tuit, and have performed a more detailed review.

stevenroose commented on bip-txhash.md:
```diff
@@ -0,0 +1,269 @@
+# Summary
+
+## OP_CHECKTXHASHVERIFY
+
```
Yeah, bare OP_CHECKTXHASHVERIFY is really efficient. CTV also adds them. It's 34 bytes output script and 0 bytes witness/scriptsig. As opposed to 34 (spk) + 33 (cb: ver + internal key) + 34 (tapscript) for taproot.

rustyrussell replied:
Ah, I forgot that OP_SUCCESSx was only a taproot thing, not a segwit thing.  Yuck!

stevenroose commented on bip-txhash.md:
```diff
@@ -0,0 +1,269 @@
+    * &nbsp;3. `TXFS_INOUT_NUMBER | TXFS_INOUT_SELECTION_ALL`
+    * &nbsp;4. `TXFS_INOUT_NUMBER | TXFS_INOUT_SELECTION_ALL`
+  * the `0x00` byte: it is set equal to `TXFS_SPECIAL_ALL`, which means "ALL" and is primarily
+    useful to emulate `SIGHASH_ALL` when `OP_TXHASH` is used in combination
```
I think the SUCCESS argument has merit, though. Also IMO it's not too much of a pain to pick one of the many SUCCESS opcodes tapscript still has to make a OP_TXHASH2 if really needed. I also don't like that user input can turn an opcode into a SUCCESS operation for the entire script. This can be tricky when collaboratively constructing scripts.

stevenroose commented on bip-txhash.md:
```diff
@@ -0,0 +1,269 @@
+summary, followed by a reference implementation of the CalculateTxHash function.
+
+* There are two special cases for the TxFieldSelector:
+  * the empty value, zero bytes long: it is set equal to `TXFS_SPECIAL_TEMPLATE`,
```
Ah sorry, one of them is a typo and should be `TXFS_SPECIAL_ALL`. Fixing.

stevenroose commented on bip-txhash.md:
```diff
@@ -0,0 +1,269 @@
+    * &nbsp;4. `TXFS_INOUT_NUMBER | TXFS_INOUT_SELECTION_ALL`
+
+* The first byte of the TxFieldSelector has its 8 bits assigned as follows, from lowest to highest:
+  * &nbsp;1. version (`TXFS_VERSION`)
+  * &nbsp;2. locktime (`TXFS_LOCKTIME`)
+  * &nbsp;3. current input index (`TXFS_CURRENT_INPUT_IDX`)
+  * &nbsp;4. current input control block (or empty) (`TXFS_CURRENT_INPUT_CONTROL_BLOCK`)
+  * &nbsp;5. current script last `OP_CODESEPARATOR` position (or 0xffffffff)
+    (`TXFS_CURRENT_INPUT_LAST_CODESEPARATOR_POS`)
+  * &nbsp;6. inputs (`TXFS_INPUTS`)
+  * &nbsp;7. outputs (`TXFS_OUTPUTS`)
```
Yeah this was annoying me a lot. It's 10x more readable than the mediawiki from before, though. Found a way to fix it.

halseth commented on bip-txhash.md:
```diff
@@ -0,0 +1,269 @@
+* The element on the stack is at least 32 bytes long, fail otherwise.
+* The first 32 bytes are interpreted as the TxHash and the remaining suffix bytes specify the TxFieldSelector.
+* If the TxFieldSelector is invalid, fail.
+* The actual TxHash of the transaction at the current input index, calculated
```
Should maybe specify that the element is not popped off the stack, or is that implicit?

stevenroose replied:
Hmm, it might be worth mentioning yeah, but I thought it was implicit as the other opcode explicitly mentions that it takes the items from the stack. It's kinda characteristic of a `-VERIFY` opcode to not touch the stack.

halseth commented on bip-txhash.md:
```diff
@@ -0,0 +1,269 @@
+    * &nbsp;4. `TXFS_INOUT_NUMBER | TXFS_INOUT_SELECTION_ALL`
+  * the `0x00` byte: it is set equal to `TXFS_SPECIAL_ALL`, which means "ALL" and is primarily
+    useful to emulate `SIGHASH_ALL` when `OP_TXHASH` is used in combination
+    with `OP_CHECKSIGFROMSTACK`.<br>Special case `TXFS_SPECIAL_TEMPLATE` is 4
```
This should be `TXFS_SPECIAL_ALL`? Maybe same as Rusty is saying.

stevenroose replied:
Fixed I think.

halseth commented on bip-txhash.md:
```diff
@@ -0,0 +1,269 @@
+summary, followed by a reference implementation of the CalculateTxHash function.
+
+* There are two special cases for the TxFieldSelector:
+  * the empty value, zero bytes long: it is set equal to `TXFS_SPECIAL_TEMPLATE`,
```
"what" is set equal to `TXFS_SPECIAL_TEMPLATE`? Maybe define what the bytes of the field selector means before the special cases.

stevenroose replied:
I think I improved this section. What I mean is "The input txfield selector is set from empty to this one, so whatever that one means".

halseth commented on bip-txhash.md:
```diff
@@ -0,0 +1,269 @@
+    * &nbsp;3. `TXFS_INOUT_NUMBER | TXFS_INOUT_SELECTION_ALL`
+    * &nbsp;4. `TXFS_INOUT_NUMBER | TXFS_INOUT_SELECTION_ALL`
+
+* The first byte of the TxFieldSelector has its 8 bits assigned as follows, from lowest to highest:
```
Found this section very hard to follow. Would it be an idea to more gently introduce an example field selector to show how it looks like (bit representation)?

stevenroose replied:
Hmm, it might be the case. In the latest version I added some example bit selectors after the written explanation. Can you see if they make sense to you?

starius commented on bip-txhash.md:
```diff
@@ -0,0 +1,269 @@
+* all in/outputs
+* the current input index
+* the leading in/outputs up to 8192
+* up to 64 individually selected in/outputs
```
Does this allow everlasting covenants (coins locked forever in a fixed set of addresses)? It is not clear from the text of BIP if this is possible and an intended use-case. IIUC, it is impossible because of chicken-and-egg problem: output script has to include a hash of itself to make an everlasting covenant which is impossible. But it is better to clarify this in BIP text explicitly, including a formal proof of why an everlasting covenant is impossible or (if it is possible) discuss use-cases and consequences.

stevenroose replied:
Yeah, this is not possible AFAIK. But I wouldn't be comfortable making that claim, if you see what people can do. Especially if we get OP_CAT and OP_TWEAKADD.

stevenroose replied:
Yeah this super naive variant is possible. It could be used for softchains (merge mining), I heard.

Roasbeef commented:
Assigned BIP 346.

murchandamus commented on bip-txhash.md:
````diff
@@ -0,0 +1,352 @@
+```
+  BIP: tbd
+  Layer: Consensus (soft fork)
+  Title: OP_TXHASH and OP_CHECKTXHASHVERIFY
+  Author: Steven Roose <steven@roose.io>
+  Comments-URI: https://github.com/bitcoin/bips/wiki/Comments:BIP-tbd
+  Status: Draft
+  Type: Standards Track
+  Created: 2023-09-03
+  License: BSD-3-Clause
````
Please incorporate the assigned number, add the README table entry, and add the Post-History header to link to the mailing list discussion or other fora where this proposal was discussed.

murchandamus reviewed (commented):
There seem to be a lot of unresolved comments in this PR, and the document appears to still be missing the Backwards Compatibility section. Please resolve existing review comments and let us know when this is ready for an editor review

0xBEEFCAF3 commented on bip-txhash.md:
```diff
@@ -0,0 +1,352 @@
+    * 8: values (`TXFS_OUTPUTS_VALUES`)
+
+* We define as follows:
+  * `TXFS_ALL = TXFS_VERSION | TXFS_LOCKTIME | TXFS_CURRENT_INPUT_IDX | TXFS_CURRENT_INPUT_CONTROL_BLOCK | TXFS_CURRENT_INPUT_LAST_CODESEPARATOR_POS | TXFS_INPUTS | TXFS_OUTPUTS | TXFS_CONTROL`
```
I think this is a typo?
```suggestion
  * `TXFS_ALL = TXFS_VERSION | TXFS_LOCKTIME | TXFS_CURRENT_INPUT_IDX | TXFS_CURRENT_INPUT_CONTROL_BLOCK | TXFS_CURRENT_INPUT_LAST_CODESEPARATOR_POS | TXFS_INPUTS_ALL | TXFS_OUTPUTS_ALL | TXFS_CONTROL`
```

ProofOfKeags commented on bip-txhash.md:
```diff
@@ -0,0 +1,352 @@
+The TxFieldSelector has the following semantics. We will give a brief conceptual
+summary, followed by a reference implementation of the CalculateTxHash function.
+
+* There are two special cases for the TxFieldSelector:
```
Can I suggest that we move the section on the special cases to be *after* the fundamental atomic selector flags? It's much easier to read this document by building up from the building blocks rather than starting with the template and having to squint to find where all the sub-properties are defined later.

I think the discussion for the optimized case follows very naturally once we understand the components here.

ProofOfKeags commented on bip-txhash.md:
````diff
@@ -0,0 +1,352 @@
+* second byte
+
+```
+ v outputs
+<-> <---------> inputs
+1 1 1 1 1 1 1 1
+| | | | | | | ^ prevouts
+| | | | | | ^ sequences
+| | | | | ^ scriptSigs
+| | | | ^ prevout scriptPubkeys
+| | | ^ prevout values
+| | ^ taproot annexes
+| ^ scriptPubkeys
+^ values
````
The "v" threw me off so I'm offering a suggestion that may help readability.

```suggestion
<-> outputs
| | <---------> inputs
1 1 1 1 1 1 1 1
| | | | | | | ^ prevouts
| | | | | | ^ sequences
| | | | | ^ scriptSigs
| | | | ^ prevout scriptPubkeys
| | | ^ prevout values
| | ^ taproot annexes
| ^ scriptPubkeys
^ values
```

stevenroose replied:
ack thanks

ProofOfKeags commented on bip-txhash.md:
````diff
@@ -0,0 +1,352 @@
+
+"only indices 0 and 2"
+```
+0 1 0 0 0 0 1 0  0 0 0 0 0 0 0 1  0 0 0 0 0 0 1 1
+| | | |                           <--------------> second idx: 3
+| | | |          <--------------> first idx: 1
+| | | | <-----> selection count: 0b10 == 2
+| | | ^ index size 0: single byte per index
+| | ^ absolute index
+| ^ individual mode
+^ don't commit the number of in/outputs
````
```suggestion
0 1 0 0 0 0 1 0  0 0 0 0 0 0 0 1  0 0 0 0 0 0 1 1
| | | |                           <-------------> second idx: 3
| | | |          <-------------> first idx: 1
| | | | <-----> selection count: 0b10 == 2
| | | ^ index size 0: single byte per index
| | ^ absolute index
| ^ individual mode
^ don't commit the number of in/outputs
```

ProofOfKeags reviewed (commented):
Did a first pass. Big Concept ACK from me. I have some notes about doc organization and plan to do another pass really drilling into the details soon.

murchandamus commented:
Hey @stevenroose, this pull request has had unaddressed review for over six months. Are you still working on this? If not, is there someone else that wants to pick this project up?

jonatack commented:
> Hey @stevenroose, this pull request has had unaddressed review for over six months. Are you still working on this? If not, is there someone else that wants to pick this project up?

pinging @stevenroose

stevenroose commented:
Hey all, sorry for the long delay. I addressed some of the above comments, and also made some changes to the BIP, together with @reardencode. Changes are the following:

- slightly modified the TEMPLATE special case to more correctly be equivalent to CTV 
- removed the 0x00 ALL special case, instead
- add a special 1-byte version of the TxFieldSelector, that maps directly into a full TxFieldSelector
  - this 1-byte txfs can directly emulate APO, APOAS and all currently existing sighashes, as outlined in the bip text
- added wording for the bip to be an upgrade path of an existing CTV deployment

murchandamus commented on bip-0346.md:
```diff
@@ -0,0 +1,438 @@
+
+This BIP proposes two new opcodes, `OP_CHECKTXHASHVERIFY`, to be activated
+as a change to the semantics of `OP_NOP4` in legacy script, segwit and tapscript;
+and `OP_TXHASH`, to be activated as a change to the semantics of `OP_SUCCESS189`
```
I double checked, OP_SUCCESS appears to be unclaimed, OP_VAULT used `OP_SUCCESS187` and `OP_SUCCESS188`.

murchandamus commented on bip-0346.md:
```diff
@@ -0,0 +1,438 @@
+## OP_CHECKTXHASHVERIFY
+
+The first new opcode, `OP_CHECKTXHASHVERIFY`, redefines the `OP_NOP4` opcode
+(`0xb3`) as a soft fork upgrade. This opcode is active in both legacy script,
+segwitv0 p2wsh and tapscript contexts.
```
I count three different contexts, so maybe:

```suggestion
(`0xb3`) as a soft fork upgrade. This opcode is active in legacy script,
segwitv0 p2wsh, and tapscript contexts.
```

murchandamus commented on bip-0346.md:
```diff
@@ -0,0 +1,438 @@
+* The element on the stack is at least 32 bytes long, NOP otherwise.
+* The first 32 bytes are interpreted as the StackTxHash and the remaining
+  suffix bytes specify the TxFieldSelector.
+* The TxFieldSelector valid, fail otherwise.
```
```suggestion
* The TxFieldSelector is valid, fail otherwise.
```

murchandamus commented on bip-0346.md:
```diff
@@ -0,0 +1,438 @@
+## OP_TXHASH
+
+The second new opcode, `OP_TXHASH`, redefines the `OP_SUCCESS189` tapscript
+opcode (`0xbd`) as a soft fork upgrade. This opcode is only active in tapscript
+context.
```
maybe "in tapscript" or "in the context of tapscript"?

```suggestion
opcode (`0xbd`) as a soft fork upgrade. This opcode is only active in tapscript.
```

murchandamus commented on bip-0346.md:
```diff
@@ -0,0 +1,438 @@
+    scriptPubkeys and amounts.
+
+    Special case `TXFS_SPECIAL_TEMPLATE` is 4 bytes long, as follows:
+    * 1: `TXFS_VERSION | TXFS_LOCKTIME | TXFS_CURRENT_INPUT_IDX`
```
How is the pipe symbol to be interpreted here?

murchandamus commented on bip-0346.md:
```diff
@@ -0,0 +1,438 @@
+    Additionally, it includes `TXFS_VERSION | TXFS_LOCKTIME | TXFS_CONTROL | TXFS_CURRENT_INPUT_TAPROOT_ANNEX | TXFS_INPUTS_SEQUENCES | TXFS_INPUTS_SCRIPTSIGS | TXFS_OUTPUTS_ALL`
+
+    These 1-byte selections allow the TxFieldSelector to emulate current
+    signature hashing modes and those defined in BIP-118:
```
```suggestion
    signature hashing modes and those defined in [BIP-118: SIGHASH_ANYPREVOUT for Taproot Scripts](bip-0118.mediawiki):
```

murchandamus commented on bip-0346.md:
```diff
@@ -0,0 +1,438 @@
+| `SINGLE|ANYPREVOUTANYSCRIPT` | `0b11000101`           |
+| `NONE|ANYPREVOUTANYSCRIPT`   | `0b11000001`           |
+
+* The first byte of the TxFieldSelector has its 8 bits assigned as follows, from lowest to highest:
```
It seems to me that we are now in the case where the TxFieldSelector is longer than 1 byte. It would perhaps be helpful to explicitly communicate that:

```suggestion
* If the TxFieldSelector is longer than 1 byte, the first byte of the TxFieldSelector has its 8 bits assigned as follows, from lowest to highest:
```

murchandamus commented on bip-0346.md:
```diff
@@ -0,0 +1,438 @@
+  `OP_TX`), this bit can be repurposed.)
+
+* The second byte will be used to indicate fields from the inputs and outputs.
+  If there is only a single byte present, no information from the inputs and
```
I’m probably missing something, but the option for the TxFieldSelector to only have a single byte here appears to be at odds with the special case above in line 83 "If the TxFieldSelector has exactly 1 byte, […]"

murchandamus commented on bip-0346.md:
```diff
@@ -0,0 +1,438 @@
+    * 8: values (`TXFS_OUTPUTS_VALUES`)
+
+* We define as follows:
+  * `TXFS_ALL = TXFS_VERSION | TXFS_LOCKTIME | TXFS_CURRENT_INPUT_IDX | TXFS_CURRENT_INPUT_CONTROL_BLOCK | TXFS_CURRENT_INPUT_LAST_CODESEPARATOR_POS | TXFS_CONTROL`
+  * `TXFS_INPUTS_ALL = TXFS_INPUTS_PREVOUTS | TXFS_INPUTS_SEQUENCES | TXFS_INPUTS_SCRIPTSIGS | TXFS_INPUTS_PREV_SCRIPTPUBKEYS | TXFS_INPUTS_PREV_VALUES | TXFS_INPUTS_TAPROOT_ANNEXES`
+  * `TXFS_OUTPUTS_ALL = TXFS_OUTPUTS_SCRIPTPUBKEYS | TXFS_OUTPUTS_VALUES`
```
Please explain the semantics of these lines. Does the pipe represent string concatenation, or some other form of concatenation here?

murchandamus commented on bip-0346.md:
```diff
@@ -0,0 +1,438 @@
+  budget ("sigops budget") introduced in BIP-0342, each TransactionHash
+  decreases the validation budget by 25. If this brings the budget below zero,
+  the script fails immediately.<br>The following considerations should be made:
+  * All fields that can be of arbitrary size are cachable as TransactionHash
```
```suggestion
  * All fields that can be of arbitrary size are cacheable as TransactionHash
```

murchandamus commented on bip-0346.md:
```diff
@@ -0,0 +1,438 @@
+* The TxFieldSelector construction would work well with a hypothetical opcode
+  `OP_TX` that allows for directly introspecting the transaction by putting the
+  fields selected on the stack instead of hashing them together.
+* The TransactionHash obtained by `OP_TXHASH` can be combined with a
+  hypothetical opcode `OP_CHECKSIGFROMSTACK` to effectively create an
```
You mean like the one proposed in BIP 348? ;)

```suggestion
* The TransactionHash obtained by `OP_TXHASH` can be combined with the
  proposed opcode `OP_CHECKSIGFROMSTACK` (see [BIP-348](bip-0348.md)) to effectively create an
```

murchandamus commented on bip-0346.md:
```diff
@@ -0,0 +1,438 @@
+
+## Comparing with some alternative proposals
+
+* This proposal strictly generalizes BIP-119's `OP_CHECKTEMPLATEVERIFY`, as the
```
```suggestion
* This proposal strictly generalizes [BIP-119](bip-0119.mediawiki)'s `OP_CHECKTEMPLATEVERIFY`, as the
```

murchandamus commented on bip-0346.md:
````diff
@@ -0,0 +1,438 @@
+```
````
```suggestion
<pre>
```

murchandamus commented on bip-0346.md:
````diff
@@ -0,0 +1,438 @@
+  Type: Standards Track
+  Created: 2023-09-03
+  License: BSD-3-Clause
+```
````
```suggestion
</pre>
```

murchandamus commented on bip-0346.md:
```diff
@@ -0,0 +1,438 @@
+  `<txfs> OP_TXHASH <pubkey> OP_CHECKSIGFROMSTACK` effectively emulates `SIGHASH_ANYPREVOUT`.
+
+
+# Detailed Specification
```
This seems to be talking about the Reference Implementation, not a Specification?

murchandamus commented on bip-0346.md:
```diff
@@ -0,0 +1,438 @@
+potentially be renamed to `OP_TEMPLATEHASH`.
+
+
+# Summary
```
Should this section perhaps be called "Specification"?

murchandamus commented on bip-0346.md:
```diff
@@ -0,0 +1,438 @@
+implementation.
+
+
+# Backwards Compatibility
```
This might also be a good place to mention the interactions with BIP 118, BIP 119, BIP 347, and BIP 348.

murchandamus reviewed (commented):
This seems to be pretty far along.
There are a few minor formatting issues which are described in line.

Please add the required Copyright section and consider renaming the "Summary" section "Specification".

instagibbs commented on README.mediawiki:
```diff
@@ -1134,6 +1134,13 @@ Those proposing changes should consider that ultimately consent may rest with th
+|-
+| [[bip-0346.mediawiki|346]]
+| Consensus (soft fork)
+| OP_TXHASH and OP_TXHASHVERIFY
```
```suggestion
| OP_TXHASH and OP_CHECKTXHASHVERIFY
```

jonatack commented on README.mediawiki:
```diff
@@ -1134,6 +1134,13 @@ Those proposing changes should consider that ultimately consent may rest with th
 | Standard
 | Draft
+|-
+| [[bip-0346.mediawiki|346]]
```
CI is failing in part because this section should be inserted between BIPs 345 and 347.

murchandamus replied:
BIP 346 TXHASH is still inserted between BIP 349 and BIP 350, while it should be between BIP 345 and BIP 347. This is one of the reasons why CI is failing. Another is the [incorrect formatting of the Author header](https://github.com/bitcoin/bips/pull/1500#discussion_r1916866915).

jonatack commented:
@stevenroose mind updating this draft per the review feedback (thanks!)

Christewart commented on bip-0346.md:
```diff
@@ -0,0 +1,438 @@
+* The element on the stack is at least 32 bytes long, NOP otherwise.
+* The first 32 bytes are interpreted as the StackTxHash and the remaining
+  suffix bytes specify the TxFieldSelector.
+* The TxFieldSelector valid, fail otherwise.
```
Still digesting this proposal so apologies if bad question, but it might make sense for this opcode to trivially succeed if the tx field selector is not well defined by this BIP? This gives us flexibility in the future with extending how tx's are hashe? 🤷‍♂️

stevenroose replied:
That's not possible. We want to push something to the stack and going from nothing to the stack to something on the stack is a breaking change. Unfortunately it's not possible to have softfork-able hooks into this opcode.

murchandamus commented:
Hey @stevenroose, this PR has had unaddressed review for over five months. The document seems close to ready for being merged in Draft status. Could you provide an update on what you expect the next steps to be here? Are you still planning to finish this? Would you like someone else to help?

jonatack reviewed (commented):
Hi @stevenroose, are you still planning to update/finish this BIP draft? Would it be worth bringing in someone to help?

Propose closing and marking as up for grabs if no reply or update by Jan 1st, 2026.

Nuhvi commented:
Not sure this is the best place to report this, but I tried to copy the `ref-impl` and use it to test the test vectors here https://github.com/Nuhvi/sake/blob/142b4fb099a4253979549193ae2dfc09ce4caf12/src/exec/sake_opcodes/op_ctv.rs#L99 and I noticed that most of the test vectors that tests the 3rd input (index 2, a taproot input) fail:
```
[
    "0:18 (10000000 #2)",
    "0:34 (ff000080 #2)",
    "0:38 (ff400040 #2)",
    "0:41 (ff800040 #2)",
    "0:44 (ffc00040 #2)",
    "0:47 (ff40003f #2)",
    "0:51 (ff80003f #2)",
    "0:55 (ffc0003f #2)",
    "0:59 (ff4000c0 #2)",
    "0:62 (ff8000c0 #2)",
    "0:65 (ffc000c0 #2)",
    "0:68 (ff4000bf #2)",
    "0:72 (ff8000bf #2)",
    "0:76 (ffc000bf #2)",
    "0:80 (ff0080 #2)",
    "0:84 (ff0140 #2)",
    "0:88 (ff0240 #2)",
    "0:92 (ff0440 #2)",
    "0:96 (ff0840 #2)",
    "0:100 (ff1040 #2)",
    "0:104 (ff2040 #2)",
    "0:108 (ff3f40 #2)",
    "0:112 (ff013f #2)",
    "0:116 (ff023f #2)",
    "0:120 (ff043f #2)",
    "0:124 (ff083f #2)",
    "0:128 (ff103f #2)",
    "0:132 (ff203f #2)",
    "0:136 (ff3f3f #2)",
    "0:140 (ff01c0 #2)",
    "0:144 (ff02c0 #2)",
    "0:148 (ff04c0 #2)",
    "0:152 (ff08c0 #2)",
    "0:156 (ff10c0 #2)",
    "0:160 (ff20c0 #2)",
    "0:164 (ff3fc0 #2)",
    "0:168 (ff01bf #2)",
    "0:172 (ff02bf #2)",
    "0:176 (ff04bf #2)",
    "0:180 (ff08bf #2)",
    "0:184 (ff10bf #2)",
    "0:188 (ff20bf #2)",
    "0:192 (ff3fbf #2)",
    "0:196 (ffff3f3f #2)",
    "0:200 (ffff4040 #2)",
    "0:203 (ff008080 #2)",
    "0:207 (ffffbfbf #2)",
    "0:211 (ffffc0c0 #2)",
    "0:214 (ffff403f #2)",
    "0:218 (ffff3f40 #2)",
    "0:221 (ffff0182 #2)",
    "0:225 (ffffc002 #2)",
    "0:228 (ffff4101420002 #2)",
    "0:232 (ffffc101c20002 #2)",
    "0:235 (ffff617f627f00 #2)",
    "0:237 (ffffe17fe27f00 #2)",
    "0:244 (ff #2)",
    "0:248 (f7 #2)",
    "0:251 (f3 #2)",
    "0:255 (fd #2)",
    "0:259 (f5 #2)",
    "0:262 (f1 #2)",
    "0:266 (ed #2)",
    "0:270 (e5 #2)",
    "0:273 (e1 #2)",
    "0:277 (cd #2)",
    "0:281 (c5 #2)",
    "0:284 (c1 #2)",
]
```

murchandamus commented:
Hi @stevenroose, thanks for working on updating your document. The PR is not ready for review or to be merged, as it is failing CI checks. 
You may have overlooked that there were still outstanding review comments that hadn’t been processed. Among other things, I described how to fix the CI issues [in my review from January 2025](https://github.com/bitcoin/bips/pull/1500#pullrequestreview-2553115630). Please process and respond to the review comments that remain unaddressed.

murchandamus reviewed (commented):
BIP 3 was deployed and the Preamble changed slightly:

murchandamus commented on bip-0346.md:
```diff
@@ -0,0 +1,444 @@
+
+`OP_TXHASH` redefines the `OP_SUCCESS189` tapscript opcode (`0xbd`) as a soft
+fork upgrade. This opcode is only active in tapscript contexts. Note that
+`OP_SUCCESS187` and `OP_SUCCESS188` are used by BIP-345 (OP_VAULT), making
```
Nit: OP_VAULT was withdrawn, so it no longer lays claim to those. However, BIP443: OP_CHECKCONTRACTVERIFY now claims OP_SUCCESS187.

reardencode reviewed (commented):
One suggestion. Otherwise LGTM.
